core/src/main/java/com/datastrato/gravitino/lock/LockManager.java

```diff
@@ -0,0 +1,119 @@
+  private final Map<NameIdentifier, LockNode> lockNodeMap = Maps.newConcurrentMap();
+  private final ThreadLocal<Stack<LockObject>> currentLocked = ThreadLocal.withInitial(Stack::new);
+
+  public LockManager() {
+    lockNodeMap.put(ROOT, new LockNode());
+  }
+
+  static class LockObject {
+    private final LockNode lockNode;
+    private final LockType lockType;
+
+    private LockObject(LockNode lockNode, LockType lockType) {
+      this.lockNode = lockNode;
+      this.lockType = lockType;
+    }
+
+    static LockObject of(LockNode lockNode, LockType lockType) {
+      return new LockObject(lockNode, lockType);
+    }
+  }
+
+  private Stack<NameIdentifier> getResourcePath(NameIdentifier identifier) {
+    Stack<NameIdentifier> nameIdentifiers = new Stack<>();
+    if (identifier == NameIdentifier.ROOT) {
+      return nameIdentifiers;
+    }
+
+    NameIdentifier ref = identifier;
+    do {
+      ref = ref.parent();
+      nameIdentifiers.push(ref);
+    } while (ref.hasParent());
+
+    return nameIdentifiers;
+  }
+
+  @VisibleForTesting
+  LockNode getOrCreateLockNode(NameIdentifier identifier) {
+    LockNode lockNode = lockNodeMap.get(identifier);
+    if (lockNode == null) {
+      synchronized (this) {
+        lockNode = lockNodeMap.get(identifier);
+        if (lockNode == null) {
+          lockNode = new LockNode();
+          lockNodeMap.put(identifier, lockNode);
+        }
+      }
+    }
+    return lockNode;
+  }
+
+  public void lockResourcePath(NameIdentifier identifier, LockType lockType) {
+    Stack<NameIdentifier> parents = getResourcePath(identifier);
+
+    try {
+      // Lock parent with read lock
+      while (!parents.isEmpty()) {
+        NameIdentifier parent = parents.pop();
+        LockNode lockNode = getOrCreateLockNode(parent);
+        lockNode.lock(LockType.READ);
+        addLock(lockNode, LockType.READ);
+      }
+
+      // Lock self with the value of `lockType`
+      LockNode node = getOrCreateLockNode(identifier);
+      node.lock(lockType);
+      addLock(node, lockType);
+    } catch (Exception e) {
+      // Unlock those that have been locked.
+      rollbackLocks();
+      LOG.error("Failed to lock resource path {}", identifier, e);
+      throw e;
+    }
+  }
+
+  public void unlockResourcePath(LockType lockType) {
+    Stack<LockObject> stack = currentLocked.get();
+    if (stack.isEmpty()) {
+      return;
+    }
+    stack.pop().lockNode.release(lockType);
+
+    while (!stack.isEmpty()) {
+      LockObject lockObject = stack.pop();
+      lockObject.lockNode.release(lockObject.lockType);
+    }
+  }
+
+  private void rollbackLocks() {
+    Stack<LockObject> locks = currentLocked.get();
+    while (!locks.isEmpty()) {
+      LockObject lockObject = locks.pop();
+      lockObject.lockNode.release(lockObject.lockType);
+    }
+  }
+
+  private void addLock(LockNode lockNode, LockType lockType) {
+    currentLocked.get().push(LockObject.of(lockNode, lockType));
+  }
+}
```

Comment:
I think the simple implementation like this is more clear.
```
public class TreeLock {
    Stack<ReentrantReadWriteLock> locks = new Stack<ReentrantReadWriteLock>();
    // [metalake, metalake.catalog, metalake.catalog.schema, metalake.catalog.schema.table]
    NameIdentifiers[] resource_names;

    TreeLock(NameIdentifier identifier) {
        this.resource_name = getResourcePath(identifier);
    }

    public void lockRead() {
        for (int i=0; i< resource_names.length-1; i++) {
            ReentrantReadWriteLock lock = LockManager.getLock(resource_names[i]);
            lock.readLock().lock();
            locks.push(lock);
        }

        ReentrantReadWriteLock lock = LockManager.getLock(resource_names[i]);
        lock.readLock().lock();
        locks.push(lock);
    }

    public void lockRead() {
        for (int i=0; i< resource_names.length-1; i++) {
            ReentrantReadWriteLock lock = LockManager.getLock(resource_names[i]);
            lock.readLock().lock();
            locks.push(lock);
        }

        ReentrantReadWriteLock lock = LockManager.getLock(resource_names[i]);
        lock.write().lock();
        locks.push(lock);
    }

    public void unlock() {
        for (int i=0; i< resource_names.length; i++) {
            ReentrantReadWriteLock lock = locks.pop();
            lock.unlock();
        }
    }
}
``

Reply:
Yeah, I think we'd better make it more simple.